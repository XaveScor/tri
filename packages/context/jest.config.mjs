import deepmerge from 'deepmerge';
import commonConfig from '../../jest.config.common.mjs';

export default deepmerge(commonConfig, {
  displayName: '@tri/context',
  rootDir: 'src',
});
