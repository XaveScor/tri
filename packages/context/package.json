{
  "name": "@widgeteria/context",
  "version": "0.0.3",
  "description": "",
  "main": "src/index.ts",
  "scripts": {
    "build": "node ../../scripts/build.mjs ./dist",
    "test": "jest"
  },
  "author": "",
  "license": "ISC",
  "type": "module",
  "dependencies": {
    "@widgeteria/hierarchy-message-bus": "^0.0.2"
  }
}
