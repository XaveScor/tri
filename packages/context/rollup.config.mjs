import Config from '../../rollup.config.mjs';

export default Config;
