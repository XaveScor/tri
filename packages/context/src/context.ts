import {
  MessageBus,
  noParentContextSymbol,
} from '@widgeteria/hierarchy-message-bus';

export class WidgeteriaContext<BaseContext, RouteArgs> {
  readonly #baseContext: BaseContext;
  readonly #routeArgs: RouteArgs;
  readonly #messageBus: MessageBus<WidgeteriaContext<BaseContext, RouteArgs>>;
  readonly #parent:
    | WidgeteriaContext<BaseContext, RouteArgs>
    | typeof noParentContextSymbol;

  static create<BaseContext, RouteArgs>(
    baseContext: BaseContext,
    routeArgs: RouteArgs,
  ) {
    return new WidgeteriaContext(baseContext, routeArgs, noParentContextSymbol);
  }

  private constructor(
    baseContext: BaseContext,
    routeArgs: RouteArgs,
    parent:
      | typeof noParentContextSymbol
      | WidgeteriaContext<BaseContext, RouteArgs>,
  ) {
    this.#baseContext = baseContext;
    this.#routeArgs = routeArgs;
    this.#parent = parent;
    this.#messageBus =
      parent === noParentContextSymbol
        ? new MessageBus<WidgeteriaContext<BaseContext, RouteArgs>>((c) => {
            return c.#getParentContext();
          })
        : parent.getMessageBus();
  }

  #getParentContext():
    | WidgeteriaContext<BaseContext, RouteArgs>
    | typeof noParentContextSymbol {
    return this.#parent;
  }

  getBaseContext() {
    return this.#baseContext;
  }

  getRouteArgs() {
    return this.#routeArgs;
  }

  getMessageBus() {
    return this.#messageBus;
  }

  createChild() {
    return new WidgeteriaContext(this.#baseContext, this.#routeArgs, this);
  }
}

type CreateWidgeteriaContextArgs<BaseContext, RouteArgs> = {
  baseContext: BaseContext;
  routeArgs: RouteArgs;
};
export function createWidgeteriaContext<BaseContext, RouteArgs>({
  baseContext,
  routeArgs,
}: CreateWidgeteriaContextArgs<BaseContext, RouteArgs>) {
  return WidgeteriaContext.create(baseContext, routeArgs);
}
