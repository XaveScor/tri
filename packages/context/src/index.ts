export {
  createWidgeteriaContext,
  createChildrenWidgeteriaContext,
  getMessageBus,
  type WidgeteriaContext,
  getWidgeteriaInnerContext,
} from './context';
