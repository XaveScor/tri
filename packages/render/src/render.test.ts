import { widgeteriaRender } from './index';
import { ChildWidget } from './fixtures/ChildWidget';
import {
  WidgeteriaAbstractRender,
  WidgeteriaAbstractWriter,
} from '@widgeteria/abstract';
import { ParentWidget } from './fixtures/ParentWidget';
import { createWidgeteriaContext } from '@widgeteria/context';

describe('render', () => {
  it('base', async () => {
    let res = '';
    const context = createWidgeteriaContext({
      baseContext: {},
      routeArgs: {},
    });
    await widgeteriaRender({
      context,
      widgetDeclaration: ChildWidget,
      args: { a: 'from args' },
      writer: new (class extends WidgeteriaAbstractWriter {
        write(chunk: string) {
          res += chunk;
        }
      })(),
      render: new (class extends WidgeteriaAbstractRender<string> {
        render(chunk: string) {
          return chunk;
        }
      })(),
    });

    expect(res).toMatchSnapshot();
  });

  it('slot', async () => {
    let res = '';
    const context = createWidgeteriaContext({
      baseContext: {},
      routeArgs: {},
    });
    await widgeteriaRender({
      context,
      widgetDeclaration: ParentWidget,
      args: { a: 'from args' },
      writer: new (class extends WidgeteriaAbstractWriter {
        write(chunk: string) {
          res += chunk;
        }
      })(),
      render: new (class extends WidgeteriaAbstractRender<string> {
        render(chunk: string) {
          return chunk;
        }
      })(),
    });

    expect(res).toMatchSnapshot();
  });
});
