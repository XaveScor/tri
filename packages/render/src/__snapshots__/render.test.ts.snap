// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render base 1`] = `"from args"`;

exports[`render slot 1`] = `"from args split hello child"`;
